### ISSUE TITLE
Borrow checking false positive on inferred closure type and borrowed fields

### ISSUE THREAD

**Issue by senioria** [rust-lang/rust#152526]
The minimal example constructed so far:

```rust
#[derive(Default)]
struct A {
    a: Vec<u32>,
    b: Vec<u32>,
}
fn main() {
    let mut a = A::default();
    let f = if true {
        |_v| vec![]
    } else {
        |_v: &Vec<u32>| vec![0_u32]
    };
    let u = f(&a.a);
    a.a = u;
    f(&a.b);
}

```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=796505788267529c272aa93965e98969))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0506]: cannot assign to `a.a` because it is borrowed
  --> src/main.rs:14:5
   |
13 |     let u = f(&a.a);
   |               ---- `a.a` is borrowed here
14 |     a.a = u;
   |     ^^^ `a.a` is assigned to here but it was already borrowed
15 |     f(&a.b);
   |     - borrow later used here

For more information about this error, try `rustc --explain E0506`.
error: could not compile `playground` (bin "playground") due to 1 previous error

```

while the following modifications wouldn't anger the borrow checker and compiles:

- Using Vec directly: `let a = vec![]`
- Removing if: `let f = |v: &Vec<u32>| vec![0_u32]`
- Fully annotating the types: `if true { |_v: &Vec<u32>| ...`
- Removing `f(&a.b)`
